feat(common): allow default currency code to be configurable

hayouung opened the pull request:
Default currency code in CurrencyPipe is currently hardcoded to USD
and is not configurable. This commit allows the default currency code
to be configurable by adding a CURRENCY_CODE injection token.

Example:
```
providers: [{ provide: CURRENCY_CODE, useValue: "GBP" }]
...
{{ 123.45 | currency }} // outputs £123.45 as opposed to always $123.45 before
```

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Hardcoded 'USD' as default currency code in CurrencyPipe and no way to change it

Issue Number: #25461


## What is the new behavior?
Allows default currency code to be configured by providing CURRENCY_CODE in a module or component

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Hayouung commented:
I messed up the author/committer by commiting under the wrong email and now I'm not quite sure how to change it back to normal. I managed to change the author but not the committer. Is there a way to change that or would it be better opening a new PR under the correct author and committer?

petebacondarwin commented:
Thanks @Hayouung. This is an interesting change. But it would be a new public API point.
I wondered if we could actually get the default currency from the current locale instead?
Currently the locale files do not contain this but I believe that it could... perhaps from this file: https://github.com/unicode-cldr/cldr-core/blob/master/supplemental/currencyData.json ?

@kara what do you think about this?

trotyl commented:
@petebacondarwin 

> I wondered if we could actually get the default currency from the current locale instead?

I believe there could be several legal currencies for one locale, for example [this "BT" one](https://github.com/unicode-cldr/cldr-core/blob/fdb81faac9c473c161d48b8a7a98028bcac61bd1/supplemental/currencyData.json#L843-L854). Let alone there can be overlap period between the currency transition.

Technically speaking the [current extraction logic to determine locale-default currency](https://github.com/angular/angular/blob/0119f46daf8f1efda00f723c5e329b0c8566fe07/tools/gulp-tasks/cldr/extract.js#L480-L483) is not correct enough, which can only get the first non-deprecated currency, even if there are several valid currencies.

petebacondarwin commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -32,6 +32,28 @@ import {InjectionToken} from '../di/injection_token';
+ *
+ * @publicApi
+ */
+export const CURRENCY_CODE = new InjectionToken<string>('CurrencyCode');
```
I would like to rename this `DEFAULT_CURRENCY_CODE` since that is what it really is - this is the default currency used if none is specified by the user of the `CurrencyPipe`.

Then the constant in `core/src/i18n/localization.ts` would need to be renamed - perhaps `USD_CURRENCY_CODE`?

Hayouung replied:
Hi @petebacondarwin, thanks for the feedback! 

I have made the appropriate changes and think I managed to squash the commits into one... hopefully.

petebacondarwin reviewed (changes requested):
I believe this is generally a good change. Just a couple of changes before I approve.

* change the name of the token
* rebase to remove the conflicts
* squash the two commits - they are really part of the same change

Also we need to get @kara and @IgorMinar to approve the public API change.

Hayouung commented:
Any way to rerun the aio_preview test? Seems like it got an error downloading something but it was working in the build before I fixed the public api guard test (which only involved reordering).

```
Error: Request failed (status: 500): CircleCI artifact download failed (Error 502 - Bad Gateway)
    at triggerWebhook.then (/home/circleci/ng/aio/scripts/create-preview.js:26:20)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Exited with code 1
```

petebacondarwin commented:
This PR never had an AIO preview

Hayouung commented:
Hi @petebacondarwin, is there anything else I need to do with this PR or do I just need to wait for the public API change to be approved?

petebacondarwin commented:
@Hayouung  - thanks for your patience. Yes we need to get the necessary approvals but the team is busy preparing for the final release of 9.0.0.

IgorMinar reviewed (changes requested):
Hi there, shouldn't the pipe use the currency of the current locale by default? I'm unclear as to why you need this override? Could you perhaps document the usecases and how to think of this vs the locale currency in the api docs?

Can you also document the default value provided by the injection token?

Thanks!

Hayouung commented:
Hi @IgorMinar, the override is required when we want to display a currency different to the app's locale currency - e.g. if I was shopping and wanted to see the price as a different currency while maintaining the same locale.

I assumed from the comments above that extracting the default locale currency code is incomplete so stuck with overriding the hardcoded USD for now. I do agree the pipe should use the locale's currency by default though.

When you say "document", does that mean in the comments of the code? (sorry, I'm very new to contributing!)

petebacondarwin commented:
OK, so I am looking at this again...

What we need here is to include the default currency "code" in our generated locale data. It isn't at the moment, only the default currency "symbol" and "display name" are included.

Once we have that then the injectable `DEFAULT_CURRENCY_CODE` defined in this PR should be set to the code found in the current locale as defined by the injectable `LOCALE_ID`.

Then I think we are in a good place. For the vast majority of users the `CurrencyPipe` will display the default currency of the current locale out-of-the-box (this is a breaking change). This seems to me to be the most intuitive scenario.

If the current locale's default currency code is not what is wanted the developer can specify a different `DEFAULT_CURRENCY_CODE` by providing their own value for this injectable. I.E. what this PR originally set out to do.

If specific currencies are wanted per use of the `CurrencyPipe` are required, then this is set in the parameters of the pipe, which is the same as the previous behaviour.

The only issue is that for many who are setting a current locale other than `en-US`, the default currency code will change, which will mean that by default the `CurrencyPipe` will stop emitting `USD`/`$` as the default currency but instead emit the default currency of the current locale.  This seems more intuitive than always emitting USD in this scenario - but it is a breaking change.

I will add a couple of commits to this PR to add this behaviour, if that is acceptable @Hayouung .

petebacondarwin committed:
feat: i18n - include currency code in locale data

hayouung committed:
feat(common): allow default currency code to be configurable

Default currency code in CurrencyPipe is currently hardcoded to USD
and is not configurable. This commit allows the default currency code
to be configurable by adding a DEFAULT_CURRENCY_CODE injection token.

Example:
```
providers: [{ provide: DEFAULT_CURRENCY_CODE, useValue: "GBP" }]
...
{{ 123.45 | currency }} // outputs £123.45 as opposed to always $123.45 before
```

Closes: #25461

petebacondarwin committed:
refactor: update locales with extracted currency codes

The locale data extraction has been modified to include the default
currency code in the generated locale data. This commit updates these
generated files accordingly.

Hayouung commented:
@googlebot I consent.

Hayouung commented:
@petebacondarwin no problem at all, appreciate the help!

IgorMinar commented on packages/core/src/application_module.ts:
```diff
@@ -42,6 +43,10 @@ export function _localeFactory(locale?: string): string {
 }
 
+export function _getDefaultCurrencyCode(localeId: string): string {
+  return getLocaleCurrencyCode(localeId) || USD_CURRENCY_CODE;
```
Could we keep the USD as a default and deprecate this behavior in v9 (via deprecations.md and updating-to-version-9.md) and a note that in v10 the default will be the default locale's currency instead? I'm concerned that changing this behavior without much notice can cause unexpected issues and considering that we are dealing here with displaying monetary values to end users, we should be very careful about rolling out this change.

The last thing we want is for an eshop in Turkey where they potentially use USD in preference to Turkish lira but localize their app using turkish locale and translations (possible, but totally unvalidated scenario!) to suddenly have their user displayed currency switched from $5 to ₺5.

IgorMinar reviewed (changes requested):
Let's resolve the issue with the timing of the breaking change, but the rest looks great! thanks

petebacondarwin committed:
docs: add deprecation info about default currency code

In v10 the default currency code will be taken from the
current locale, rather than simply defaulting to `USD`.

petebacondarwin commented:
@IgorMinar - I have removed the last commit (and moved it to #34724) so there is no breaking change in this PR; and I have added a commit to document the deprecation. PTAL!

petebacondarwin committed:
test: update tests for new locale data

IgorMinar commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -143,10 +143,28 @@ export class PercentPipe implements PipeTransform {
+ * 
+ * The default currency code is currently always `USD` but this is deprecated from v9.
+ * 
+ * **In v10 the default currency code will be taken from the current locale.**
```
we should clarify that this is the application locale and not the locale provided by the browser. consider changing this to "current locale as configured via the https://angular.io/guide/i18n#setting-up-the-locale-of-your-app"

IgorMinar commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -143,10 +143,28 @@ export class PercentPipe implements PipeTransform {
+ * 
+ * **In v10 the default currency code will be taken from the current locale.**
+ * 
+ * If you need the previous behaviour then provide it explicitly:
```
behaviour -> behavior
"provide it" -> "set the default currency explicitly via a DI provider in your application NgModule:"

IgorMinar commented on aio/content/guide/updating-to-version-9.md:
```diff
@@ -35,6 +35,7 @@ See our [template type-checking guide](guide/template-typecheck) for more inform
 | API                                                                     | Replacement                          | Notes |
 | ------------------------------------------------------------------------| ------------------------------------ | ----- |
 | [`entryComponents`](api/core/NgModule#entryComponents)                  | none                                 | See [`entryComponents`](guide/deprecations#entryComponents) |
+| [`CurrencyPipe` - `DEFAULT_CURRENCY_CODE`](api/common/CurrencyPipe#currency-code-deprecation)| `{provide: DEFAULT_CURRENCY_CODE, useValue: 'USD'}` | From v10 the default code will be extracted from the current locale, rather than `USD`. |
```
I spoke with Stephen about this and we agreed that we should give people N+2 grace period in this case. So let's change v10 to v11. thanks!

IgorMinar reviewed (approved):
a few more changes requested but otherwise lgtm. thanks!

IgorMinar commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -143,10 +143,28 @@ export class PercentPipe implements PipeTransform {
  *
+ * {@a currency-code-deprecation}
+ * <div class="alert is-helpful">
+ * 
```
@petebacondarwin  I noticed that there is a lot of trailing whitespace chars in this file. Can you please strip them and configure your editor to strip those automatically? thanks!

petebacondarwin replied:
Strange - my VS Code used to remove whitespace. Perhaps an update to it turned it off. I have re-enabled it now. Thanks

IgorMinar commented:
I kicked off a presubmit for this: https://test.corp.google.com/ui#id=OCL:289139962:BASE:289140002:1578687403141:49549c96 

the result will be obsolete once Pete updates this PR, but since his changes touch only MD files and comments, I don't think that we need another one, so Pete you could request merge-assistance and point to this presubmit to expedite the merge.

IgorMinar commented:
hmm.. the presubmit fails because internal apps call the constructor explicitly and since the constructor signature has changed with an additional argument the code no longer compiles.

petebacondarwin commented:
If it is not a big job then we should update G3. Slightly more concerning though is why these apps are calling the constructor directly??

petebacondarwin committed:
fixup! docs: add deprecation info about default currency code

AndrewKushnir commented:
@petebacondarwin @IgorMinar I looked at a couple placed in g3 and it looks like CurrencyPipe is used as a utility function to convert values in TS code. It also looks like we should run a Global TAP, since there might be more instances that we may need to cleanup, so it might take time. Also, it's a bit unclear to me how to update apps code now to make it compatible with upcoming changes (we can not just add one more argument - it'd also be TS error I think).

Probably the easiest way would be to make the second argument optional in v9 (so we can also update g3 by adding extra argument to `CurrencyPipe` constructor) and make it required in v10 (if needed).

Thank you.

petebacondarwin commented:
OK, I can make it optional over the weekend.

petebacondarwin committed:
fixup! feat(common): allow default currency code to be configurable

petebacondarwin commented:
@AndrewKushnir - the constructor parameter is now optional. It will default to `USD` if not provided, which is basically the same as the previous behaviour, so no breaking change. PTAL.

IgorMinar commented:
merge-assistance: global approval

atscott commented:
@petebacondarwin it looks like there's a conflict with the patch branch. I'm switching this to master only. Could you create a new PR that targets patch only?
